### PR TITLE
Trust wdl

### PR DESCRIPTION
**Pull request by asmariyaz23**
- Run Trust4 
- handle both bam and fastq data

### CODE REVIEW

File: workflows/trust4/trust4.wdl [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        String barcode_range = "0,15,+"
+        # path to the barcode whitelist
+        File? barcode_whitelist
+        # UMI Fastq pattern
+        String umi_fastq_pattern = "_S\*_L*_R1_001.fastq.gz"
+        # start, end(-1 for length-1), strand in a UMI is the true UMI 
+        String umi_range = "16,-1,+"
+        # If input_bam; provide bam field for UMI
+        String? umi_bam_field
+        # path to bam file
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:49]: Please add one blank line here.

File: workflows/trust4/trust4.wdl [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                except CalledProcessError:
+                    if not os.path.exists(target):
+                        os.mkdir(target)
+                    call_args = ['strato', 'cp', '--backend', '~{backend}', '-m', directory + '/~{sample_id}*' , target + '/']
+                    print(' '.join(call_args))
+                    check_call(call_args)
+
+                fastq_dirs.append(target)
+
+        if '~{input_bam}':
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:212]: ~{input_bam} itself is a BAM file, there is no need to copy anything from bucket.

File: workflows/trust4/trust4.wdl [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            call_args.extend(['--mateIdSuffixLen', '~{mateIdSuffixLen}'])
+        if '~{noExtraction}':
+            call_args.append('--noExtraction')
+        if '~{repseq}':
+            call_args.extend(['--repseq', '~{repseq}'])
+        if '~{outputReadAssignment}':
+            call_args.append('--outputReadAssignment')
+
+        if '~{barcode_range}':
+            call_args.extend(['--barcode_range', ' '.join(~{barcode_range}'.split(','))])
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:239]: To be safe, please make each parameter as a separate item in the list.

For example `--barcodeRange 0 15 +` should be ['--barcodeRange', '0', '15', '+'].

File: workflows/trust4/trust4.wdl [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        if '~{umi_range}':
+            call_args.extend(['--umi_range', ' '.join(~{umi_range}'.split(','))])
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:241]: Please check the right TRUST4 command.

File: workflows/trust4/trust4.wdl [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                call_args.extend(['--abnormalUnmapFlag', '~{bam_abnormal_unmap_flag}'])
+            if '~{umi_bam_field}':
+                call_args.extend(['--UMI', '~{umi_bam_field}'])   
+            call_args.extend(['-b', '~{input_bam}']
+
+        if '~{input_fastqs_directories}':
+            if '~{pe_read1_fastq_pattern}' and '~{pe_read2_fastq_pattern}':
+                for pe_fastq_dir in fastq_dirs:
+                    call_args.extend(['-1', Path(pe_fastq_dir, ~{pe_read1_fastq_pattern}),
+                                      '-2', Path(pe_fastq_dir, ~{pe_read1_fastq_pattern})])
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:256]: One typo here.

File: workflows/trust4/trust4.wdl [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+        # Number of cpus per trust4 job
+        Int num_cpu = 8
+        # Memory string
+        String memory = "32G"
+        # Disk space in GB
+        Int disk_space = 200
+
+        # R2 Fastq pattern SE data
+        String se_fastq_pattern = "_S\*_L*_R2_001.fastq.gz"
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:27]: No backslash

File: workflows/trust4/trust4.wdl [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        String? read2_range
+
+        # Barcode Fastq pattern
+        String barcode_fastq_pattern = "_S*_L*_R1_001.fastq.gz"
+        # start, end(-1 for length-1), strand in a barcode is the true barcode
+        String barcode_range = "0,15,+"
+        # path to the barcode whitelist
+        File? barcode_whitelist
+        # UMI Fastq pattern
+        String umi_fastq_pattern = "_S\*_L*_R1_001.fastq.gz"
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:44]: No backslash

File: workflows/trust4/trust4.wdl [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                if '~{read2_range}':
+                    call_args.extend(['--read2Range', ' '.join('~{read2_range}'.split(','))])                      
+            else:
+                for se_fastq_dir in fastq_dirs:
+                    call_args.extend(['-u', Path(se_fastq_dir, se_fastq_pattern)])
+            
+            if '~{read1_range}':
+                call_args.extend(['--read1Range', ' '.join('~{read1_range}'.split(','))])
+
+            for barcode_fastq_dir in fastq_dirs:
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:266]: Please add a test if the barcode_pattern is null

File: workflows/trust4/trust4.wdl [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+                call_args.extend(['--barcode', Path(barcode_fastq_dir, ~{barcode_fastq_pattern})])
+
+            for umi_fastq_dir in fastq_dirs:
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:269]: Please add a test if the umi_pattern is null

**Review by bli25 — CHANGES_REQUESTED** [review]

File: workflows/trust4/trust4.wdl [hunk 10]
```diff
@@ -263,17 +257,19 @@ task run_trust4 {
             if '~{read1_range}':
                 call_args.extend(['--read1Range', ' '.join('~{read1_range}'.split(','))])
 
-            for barcode_fastq_dir in fastq_dirs:
-                call_args.extend(['--barcode', Path(barcode_fastq_dir, ~{barcode_fastq_pattern})])
+            if ~{barcode_fastq_pattern} is not None:
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:260]: '~{barcode_fastq_pattern}' != ''

File: workflows/trust4/trust4.wdl [hunk 11]
```diff
@@ -263,17 +257,19 @@ task run_trust4 {
...
+                for barcode_fastq_dir in fastq_dirs:
+                    call_args.extend(['--barcode', Path(barcode_fastq_dir, '~{barcode_fastq_pattern}')])
 
-            for umi_fastq_dir in fastq_dirs: 
-                call_args.extend(['--UMI', Path(umi_fastq_dir, ~{umi_fastq_pattern})])
+            if ~{umi_fastq_pattern} is not None:
```
Comment (bli25) [inline workflows/trust4/trust4.wdl:264]: '~{umi_fastq_pattern}' is not None

**Review by bli25 — CHANGES_REQUESTED** [review]

**Review by bli25 — APPROVED** [review]
